scanner.go

```diff
@@ -250,95 +238,136 @@ func (s *scannerV2) Next() (*hrpc.ScanResponseV2, error) {
-	var (
-		rpc *hrpc.Scan
-		err error
-	)
-
+func (s *scanner) makeScanRequest() (*hrpc.Scan, error) {
 	if s.isRegionScannerClosed() {
 		// preserve ScanStatsID
 		opts := append(s.rpc.Options(), hrpc.ScanStatsID(s.rpc.ScanStatsID()))
 
 		// open a new region scan to scan on a new region
-		rpc, err = hrpc.NewScanRange(
+		return hrpc.NewScanRange(
 			s.rpc.Context(),
 			s.rpc.Table(),
 			s.startRow,
 			s.rpc.StopRow(),
 			opts...)
-	} else {
-		// continuing to scan current region
-		opts := []func(hrpc.Call) error{
-			hrpc.ScannerID(s.curRegionScannerID),
-			hrpc.NumberOfRows(s.rpc.NumberOfRows()),
-			hrpc.Priority(s.rpc.Priority()),
-			hrpc.RenewInterval(s.rpc.RenewInterval()),
-			// preserve ScanStatsID
-			hrpc.ScanStatsID(s.rpc.ScanStatsID()),
-			hrpc.WithScanStatsHandler(s.rpc.ScanStatsHandler()),
-		}
-		if s.rpc.TrackScanMetrics() {
-			opts = append(opts, hrpc.TrackScanMetrics())
-		}
-		rpc, err = hrpc.NewScanRange(s.rpc.Context(),
-			s.rpc.Table(),
-			s.startRow,
-			nil,
-			opts...,
-		)
 	}
-	if err != nil {
-		return nil, nil, nil, err
+
+	// continuing to scan current region
+	opts := []func(hrpc.Call) error{
+		hrpc.ScannerID(s.curRegionScannerID),
+		hrpc.NumberOfRows(s.rpc.NumberOfRows()),
+		hrpc.Priority(s.rpc.Priority()),
+		hrpc.RenewInterval(s.rpc.RenewInterval()),
+		// preserve ScanStatsID
+		hrpc.ScanStatsID(s.rpc.ScanStatsID()),
+		hrpc.WithScanStatsHandler(s.rpc.ScanStatsHandler()),
+	}
+	if s.rpc.TrackScanMetrics() {
+		opts = append(opts, hrpc.TrackScanMetrics())
 	}
 
+	return hrpc.NewScanRange(s.rpc.Context(),
+		s.rpc.Table(),
+		s.startRow,
+		nil,
+		opts...,
+	)
+}
+
+// request sends the next scan request to HBase and calls update() to
+// prepare for the next request.
+func (s *scanner) request() (*hrpc.ScanResponseV2, error) {
+	rpc, err := s.makeScanRequest()
+	if err != nil {
+		s.Close()
+		return nil, err
+	}
 	res, err := s.SendRPC(rpc)
 	if err != nil {
-		return nil, nil, nil, err
+		s.Close()
+		return nil, err
 	}
 	scanres, ok := res.(*pb.ScanResponse)
 	if !ok {
-		return nil, nil, nil, errors.New("got non-ScanResponse for scan request")
+		s.Close()
+		return nil, errors.New("got non-ScanResponse for scan request")
 	}
-	return rpc.Response, scanres, rpc.Region(), nil
+
+	if s.rpc.TrackScanMetrics() && scanres.ScanMetrics != nil {
+		metrics := scanres.ScanMetrics.GetMetrics()
+		for _, m := range metrics {
+			s.scanMetrics[m.GetName()] += m.GetValue()
+		}
+	}
+
+	s.update(scanres, rpc.Region())
+
+	return rpc.Response, nil
 }
 
-// update updates the scanner for the next scan request
+// update updates the scanner for the next scan request or closes it.
 func (s *scanner) update(resp *pb.ScanResponse, region hrpc.RegionInfo) {
-	if s.isRegionScannerClosed() && resp.ScannerId != nil {
-		s.openRegionScanner(resp.GetScannerId())
+	// Either there are more results in this region, more results in
```

Comment:
This is a behaviour change in the refactored version:

* we used to check for closure before checking for more results.
* I think Close could be called asynchronously by a client
* in that case, we would then recall openRegionScanner. 

There also is potentially some racyness with setting curRegionScannerID:
* if we close, it sets it to noScannerID,
* but if we are racing in this block, we could set it here instead

Finally, isDone also checks for closure, which is now folded into this block from what I understand

I don't know if either of these things are impactful, actually the first one sounds a bit more like a bug. I'm wondering if you thought about these cases.